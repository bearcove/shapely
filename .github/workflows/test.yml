name: Tests

on:
  push:
    branches: [main]
  pull_request:
    branches: [main]
  merge_group:

jobs:
  code-quality:
    runs-on: depot-ubuntu-24.04-arm-16

    container:
      image: ghcr.io/facet-rs/facet-ci:latest
    steps:
      - uses: actions/checkout@v4

      - uses: Swatinem/rust-cache@v2

      - name: ✨ Run code quality checks
        shell: bash
        run: |
          just code-quality

  test:
    runs-on: depot-ubuntu-24.04-arm-16

    container:
      image: ghcr.io/facet-rs/facet-ci:latest
    steps:
      - uses: actions/checkout@v4

      - uses: Swatinem/rust-cache@v2

      - name: ✨ Run tests
        shell: bash
        run: |
          just test

  nostd:
    runs-on: depot-ubuntu-24.04-arm-16

    container:
      image: ghcr.io/facet-rs/facet-ci:latest
    steps:
      - uses: actions/checkout@v4

      - uses: Swatinem/rust-cache@v2

      - name: ✨ Run nostd tests
        shell: bash
        run: |
          just nostd

  doc-tests:
    runs-on: depot-ubuntu-24.04-arm-16

    container:
      image: ghcr.io/facet-rs/facet-ci:latest
    steps:
      - uses: actions/checkout@v4

      - uses: Swatinem/rust-cache@v2

      - name: ✨ Run doc tests
        shell: bash
        run: |
          just doc-tests

  miri:
    runs-on: depot-ubuntu-24.04-arm-32

    container:
      image: ghcr.io/facet-rs/facet-ci:latest-miri
    steps:
      - uses: actions/checkout@v4

      - uses: Swatinem/rust-cache@v2

      - name: ✨ Run miri
        shell: bash
        run: |
          export CI=true
          just miri

  msrv:
    runs-on: depot-ubuntu-24.04-arm-16

    container:
      image: ghcr.io/facet-rs/facet-ci:latest
    steps:
      - uses: actions/checkout@v4

      - uses: Swatinem/rust-cache@v2

      - uses: taiki-e/install-action@cargo-hack

      - name: ✨ Check MSRV
        shell: bash
        run: |
          just msrv

  minimal-versions:
    runs-on: depot-ubuntu-24.04-arm-16

    container:
      image: ghcr.io/facet-rs/facet-ci:latest
    steps:
      - uses: actions/checkout@v4

      - uses: Swatinem/rust-cache@v2

      - name: ✨ Check minimal versions
        shell: bash
        run: |
          # Install nightly for minimal-versions feature
          rustup toolchain install nightly
          # Generate lockfile with minimal versions
          cargo +nightly generate-lockfile -Z minimal-versions
          # Check with stable using minimal versions
          cargo check --workspace --all-features --locked --keep-going

  docs:
    runs-on: depot-ubuntu-24.04-arm-16

    container:
      image: ghcr.io/facet-rs/facet-ci:latest
    steps:
      - uses: actions/checkout@v4

      - uses: Swatinem/rust-cache@v2

      - name: ✨ Check documentation
        shell: bash
        env:
          RUSTDOCFLAGS: -D warnings
        run: |
          just docs

  lockfile:
    runs-on: depot-ubuntu-24.04-arm-16

    container:
      image: ghcr.io/facet-rs/facet-ci:latest
    steps:
      - uses: actions/checkout@v4

      - uses: Swatinem/rust-cache@v2

      - name: ✨ Check lockfile is updated
        shell: bash
        run: |
          cargo update --workspace --locked

  clippy:
    runs-on: depot-ubuntu-24.04-arm-16

    container:
      image: ghcr.io/facet-rs/facet-ci:latest
    permissions:
      security-events: write # to upload sarif results
    steps:
      - uses: actions/checkout@v4

      - uses: Swatinem/rust-cache@v2

      - name: Install SARIF tools
        shell: bash
        run: |
          cargo install clippy-sarif --locked
          cargo install sarif-fmt --locked

      - name: ✨ Run clippy with SARIF output
        shell: bash
        run: |
          cargo clippy --workspace --all-features --all-targets --message-format=json | clippy-sarif | tee clippy-results.sarif | sarif-fmt
        continue-on-error: true

      - name: Upload SARIF results
        uses: github/codeql-action/upload-sarif@v3
        with:
          sarif_file: clippy-results.sarif
          wait-for-processing: true

      - name: Report status
        shell: bash
        run: |
          cargo clippy --workspace --all-features --all-targets --keep-going -- -D warnings --allow deprecated

  coverage:
    runs-on: depot-ubuntu-24.04-arm-32

    container:
      image: ghcr.io/facet-rs/facet-ci:latest
    steps:
      - uses: actions/checkout@v4

      - uses: Swatinem/rust-cache@v2

      - name: ✨ Gather coverage
        shell: bash
        run: |
          cargo tarpaulin --output-dir coverage --out lcov

      - name: ✨ Publish to Coveralls
        uses: coverallsapp/github-action@master
        with:
          github-token: ${{ secrets.GITHUB_TOKEN }}
