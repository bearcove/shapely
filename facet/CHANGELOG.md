# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.10](https://github.com/facet-rs/facet/compare/facet-v0.1.9...facet-v0.1.10) - 2025-04-10

### Fixed

- typo in README.md

### Other

- Merge branch 'main' into from-ptr
- Update doc tests, too
- peek & poke are not reexported by the main crate
- Make tests pass on stable for now...
- Add an empty test (failing on rust stable)
- Merge branch 'main' into patch-1
- Fixed doc tests
- Merge facet-opaque, facet-spez, facet-types and facet-trait back into facet-core, to allow implementing Facet for Shape

### Other

- Removed `Facet::ARCHETYPE` in favor of new `SpezEmpty` type wrapper

## [0.1.9](https://github.com/facet-rs/facet/compare/facet-v0.1.8...facet-v0.1.9) - 2025-04-10

### Other

- middle ground
- and some spacing
- ok 30
- ok no ems
- image height
- use logo only for readmes
- logo v2

## [0.1.8](https://github.com/facet-rs/facet/compare/facet-v0.1.7...facet-v0.1.8) - 2025-04-10

### Fixed

- fix readmes

### Other

- remove spacing
- no height
- Update Readmes with logos.

## [0.1.7](https://github.com/facet-rs/facet/compare/facet-v0.1.6...facet-v0.1.7) - 2025-04-10

### Other

- doctests
- Link facet-args
- show off CLI parsing example
- shaep -> shape

## [0.1.6](https://github.com/facet-rs/facet/compare/facet-v0.1.5...facet-v0.1.6) - 2025-04-10

### Other

- Am excited
- Show it's actually runtime
- Specialization demo re: pretty
- Unify top-level README
- WIP readmes
- Fix link

## [0.1.5](https://github.com/facet-rs/facet/compare/facet-v0.1.4...facet-v0.1.5) - 2025-04-09

### Other

- upgrades
- Fix YAML link in README

## [0.1.4](https://github.com/facet-rs/facet/compare/facet-v0.1.3...facet-v0.1.4) - 2025-04-09

### Other

- DUMMY => ARCHETYPE

## [0.1.3](https://github.com/facet-rs/facet/compare/facet-v0.1.2...facet-v0.1.3) - 2025-04-08

### Other

- More nostd
- nostd facet-trait
- facet-types is now no_std friendly :)
- Less experimental now
- woo almost everything is non-exhaustive

## [0.1.2](https://github.com/facet-rs/facet/compare/facet-v0.1.1...facet-v0.1.2) - 2025-04-08

### Other

- preparing for json tuples
- Add support for sensitive fields
- not using namespace runners for now
- Fix links to other repos

## [3.1.1](https://github.com/facet-rs/facet/compare/facet-v3.1.0...facet-v3.1.1) - 2025-04-05

### Fixed

- fix errors now
- fix miri/memory problems
- fix derive probably

### Other

- Add .envrc to try removing rebuilds
- Facet is unsafe
- Fix tests etc.
- woo everything builds
- getting there
- The great split
- start fixing doc tests
- Cool, the hacking guide is in
- 29 tests passed aw yiss
- clone stuff
- clone in place => clone into
- mhmh
- mhhh getting somewhere
- fun
- bigger and bigger
- switch to btparse
- time to fix those tests
- mhmhmh
- color backtrace in tests please?
- tests are made to fail
- mhmh tests are failing huh
- maps, slowlyl
- peeking a lotta things
- more vec stuff
- okay, debug and default, it's something
- mhh we regressed
- Uhhh slices work?
- well this weirdly works?
- mhkay
- traits tests look better
- more spez is going well
- mh
- mhhhmhh
- mhhhhh it's probably the uninit thing, ngl
- I'm confused now
- uhhhh
- uhhh what
- weird
- more tests
- More tests
- rename spez to traits
- mhmhmh
- Uncomment a bunch of tests
- Unreasonably happy with that tbh
- nice nice
- Use spez-like ideas to set Debug if it's set on the type
- Well that's not really const
- yessssss
- getting somewhere maybe? but only in macros, they weren't joking.
- mhhh
- ahhhhh
- alright, will this work?
- Mhh doesn't work yet
- Well the tests do pass
- welp
- don't compare strings
- uncomment some derives
- so far so god
- new structure works
- introduce init_in_place_with_capacity
- Rname more things for more consistency
- Move tests to facet proper
- whoa hey down to 111 errors
- Innards => Def

## [3.1.0](https://github.com/facet-rs/facet/compare/facet-v3.0.0...facet-v3.1.0) - 2025-03-31

### Other

- Fill in missing docs
- Document more stuff
- Improve naming
- More notes
- Fix all markdown links
- More docs
- unfuck docs
- Mhh it derives _something_
- Tuple layouts
- more tuple support
- pre-commit hook with cargo-fmt
- More complex derives work
- Real-life derive, which fails
- Support more attribute
- more derive
- shill for namespace, closes #36
- just pre-commit, just fmt
- set up cargo-husky to run cargo fmt
- specific toolchains, reformat code

## [3.0.0](https://github.com/facet-rs/facet/compare/facet-v2.0.1...facet-v3.0.0) - 2025-03-11

### Other

- Fix doc tests in README.md
- Document how to write your deserializer a little better

## [2.0.0](https://github.com/facet-rs/facet/compare/facet-v1.0.0...facet-v2.0.0) - 2025-03-11

### Other

- Change Shape.name from static str to NameFn
- Looking good!
- clean up drop impls
- Stability notes
- link to 'free of syn' campaign
- Make derive unconditional, closes #8
- Get rid of insta, closes #10
- Ensure no syn dependency (and badge about it), closes #9
- Introduce core crate
- Get rid of debug/display, closes #4
- Start implementing transparent
