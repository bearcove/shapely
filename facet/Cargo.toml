[package]
name = "facet"
version = "0.2.0"
authors.workspace = true
edition.workspace = true
rust-version.workspace = true
license.workspace = true
repository.workspace = true
description = "Re-exports the Facet trait and derive macros, along with Shape and all related types"
keywords = ["facet", "introspection", "reflection"]
categories = ["development-tools"]

[dependencies]
facet-core = { path = "../facet-core", version = "0.5.3", default-features = false }
facet-derive = { path = "../facet-derive", version = "0.1.20" }
facet-reflect = { path = "../facet-reflect", version = "0.6.2", optional = true }

[features]
reflect = ["dep:facet-reflect"]   # Enables reflection via Peek and Poke types
testfeat = []                     # Does nothing, only used for tests
std = ["facet-core/std", "alloc"] # Uses libstd and alloc
alloc = ["facet-core/alloc"]      # Enables alloc support
default = ["std"]

[dev-dependencies]
cargo-husky = { version = "1.5.0", default-features = false, features = ["user-hooks"] }
eyre = { version = "0.6.12", default-features = false }
facet-reflect = { path = "../facet-reflect", version = "0.6.2" }
