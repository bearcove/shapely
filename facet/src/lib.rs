#![cfg_attr(not(feature = "std"), no_std)]
#![warn(missing_docs)]
#![warn(clippy::std_instead_of_core)]
#![warn(clippy::std_instead_of_alloc)]
#![doc = include_str!("../README.md")]
#![cfg_attr(docsrs, feature(doc_cfg))]
#![cfg_attr(docsrs, feature(builtin_syntax))]
#![cfg_attr(docsrs, feature(prelude_import))]
#![cfg_attr(docsrs, allow(internal_features))]

#[cfg(docsrs)]
pub mod sample_generated_code;

pub use facet_core::*;

pub use facet_derive::*;

#[cfg(feature = "reflect")]
pub use facet_reflect::*;

pub mod hacking;
