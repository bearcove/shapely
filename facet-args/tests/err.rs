use facet::Facet;
use facet_testhelpers::test;

#[test]
fn test_error_non_struct_type_not_supported() {
    #[derive(Facet, Debug)]
    #[repr(u8)]
    #[allow(dead_code)]
    enum Args {
        Something,
        Else,
    }
    let args: Result<Args, _> = facet_args::from_slice(&["error", "wrong", "type"]);
    let err = args.unwrap_err();
    insta::assert_snapshot!(err);
    // Args error: Expected struct type
}

#[test]
fn test_error_missing_value_for_argument() {
    #[derive(Facet, Debug)]
    struct Args {
        #[facet(named, short = 'j')]
        concurrency: usize,
    }
    let args: Result<Args, _> = facet_args::from_slice(&["--concurrency"]);
    let err = args.unwrap_err();
    insta::assert_snapshot!(err);
    // assert_eq!(
    //     err.message(),
    //     "Args error: expected value after argument `concurrency`"
    // );
}

#[test]
fn test_error_missing_value_for_argument_short_missed() {
    #[derive(Facet, Debug)]
    struct Args {
        #[facet(named, short = 'j')]
        concurrency: usize,
        #[facet(named, short = 'v')]
        verbose: bool,
    }
    let args: Result<Args, _> = facet_args::from_slice(&["-j", "-v"]);
    let err = args.unwrap_err();
    insta::assert_snapshot!(err);
    // assert_eq!(
    //     err.message(),
    //     "Args error: expected value after argument `j`"
    // );
}

#[test]
fn test_error_missing_value_for_argument_short_eof() {
    #[derive(Facet, Debug)]
    struct Args {
        #[facet(named, short = 'j')]
        concurrency: usize,
    }
    let args: Result<Args, _> = facet_args::from_slice(&["-j"]);
    let err = args.unwrap_err();
    insta::assert_snapshot!(err);
    // assert_eq!(
    //     err.message(),
    //     "Args error: expected value after argument `j`"
    // );
}

#[test]
fn test_error_unknown_argument() {
    #[derive(Facet, Debug)]
    struct Args {
        #[facet(named, short = 'j')]
        concurrency: usize,
    }
    let args: Result<Args, _> = facet_args::from_slice(&["--c0ncurrency"]);
    let err = args.unwrap_err();
    insta::assert_snapshot!(err);
    // assert_eq!(err.message(), "Args error: Unknown argument `c0ncurrency`");
}
