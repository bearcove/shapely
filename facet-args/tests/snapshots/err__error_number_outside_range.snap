---
source: facet-args/tests/err.rs
expression: err
---
[31mError:[0m 
   [38;5;246mâ•­[0m[38;5;246mâ”€[0m[38;5;246m[[0m args:1:9 [38;5;246m][0m
   [38;5;246mâ”‚[0m
 [38;5;246m1 â”‚[0m [38;5;249m-[0m[38;5;249m-[0m[38;5;249ms[0m[38;5;249mm[0m[38;5;249ma[0m[38;5;249ml[0m[38;5;249ml[0m[38;5;249m [0m[31m1[0m[31m0[0m[31m0[0m[31m0[0m
 [38;5;240m  â”‚[0m         [31mâ”€[0m[31mâ”€[0m[31mâ”¬[0m[31mâ”€[0m  
 [38;5;240m  â”‚[0m           [31mâ•°[0m[31mâ”€[0m[31mâ”€[0m[31mâ”€[0m While trying to put [32mu64[39m into a [34mu8[39m: [31mconversion from u64 failed[39m
[38;5;246mâ”€â”€â”€â•¯[0m
